Structured data/labeling tools (#14)
ml6team/quick-tips: structured_data/2021_06_07_timeseries_labeling_tools/README.md

# Time series labeling tools
## Goal

> This document tries to give an overview of tools that can be used for data labelling, specifically for time series.

Author: all set to labeling

> Do take care of scopes! Those are a legacy layer on top of IAM. You need to give your CE instance the following settings:
> - Cloud API access scopes
> - Allow full access to all Cloud APIs

By default an SQLite is used in the backend. 
If this would be too limiting, a PostgreSQL database can be used as well (https://labelstud.io/guide/storedata.html). 

In Label Studio users and projects can be created. 
Within these projects different *label tasks* can be created. 
Label tasks are time series segments that need to be labeled.
Once the operator is happy with the labelling, it can be submitted.

#### Import
To import data, a data template template needs to be made. 
This template is basically the configuration of how the data looks like (which columns are important), but also how the data can be viewed during labelling.
This allows for a very flexible import system.
More info on these data templates: [https://labelstud.io/templates/time_series.html](https://labelstud.io/templates/time_series.html)

The data needs to be formatted in wide format, which for certain cases might lead to sparse datasets.

> On top of that, Label Studio seems to set missing values (~ sparse data) to 0, complicating importing data from a single source, for data that was not taken at the same time.

Reviewer: ```suggestion
On top of that, Label Studio seems to set missing values to 0. This complicates the import when the values in different columns are sampled at different frequencies.
```

To avoid this, multiple different files can be used.

The tool can use different data sources. For an overview check: [https://labelstud.io/guide/storedata.html](https://labelstud.io/guide/storedata.html).
As an example, Label Studio can directly connect to a GCP (storage bucket): [https://labelstud.io/guide/storage.html#Google-Cloud-Storage](https://labelstud.io/guide/storage.html#Google-Cloud-Storage).

> Unfortunately this did not seem to work outside of the VPC. This is probably due to a rights issue.

Reviewer: ```suggestion
Out of the box, this did not seem to work outside of the VPC. This is probably due to incorrect roles/permissions.
```

> Data can be imported using the GUI or using an API.

Reviewer: ```suggestion
Data can also be imported via a GUI or [using the API](https://labelstud.io/guide/tasks.html#Import-data-using-the-API).
```

Files up to 200Mb are supposed to work, although it started failing above 50Mb using the GUI.
Files that are bigger suffer from a *validation error*.

> Label Studio has an API that can be used to pump data in.  ([https://labelstud.io/guide/tasks.html#Import-data-using-the-API](https://labelstud.io/guide/tasks.html#Import-data-using-the-API))

Reviewer: ```suggestion
```

Some other tips:
- Data should be sorted on time
- Missing data: data will be set to 0 (flat line).

> - Corrupt data (e.g. strings): time series just stop at the first time point it encountered a string. So even if the data afterwards is okay. There is simply no data to been seen (empty time series plot).

Reviewer: ```suggestion
- Corrupt data (e.g. strings): the data is imported correctly **until the first occurrence of a corrupt value**. Even with a single corrupt datapoint, this could mess up your whole dataset.
```

> So be a bit careful. A small anomaly might destroy the rest of the time series. Also the sparse data part is something to keep an eye on.

Reviewer: ```suggestion
```

#### Export

> Data can be exported in different ways, JSON making most sense. There is also an API, which allows to automate the export process + parsing + dumping towards e.g. GCP.

Reviewer: Do you mean a "data export", or is this about exporting the labels themselves?

Author: data and labels. fixed

[https://labelstud.io/templates/time_series.html#Output-format-example](https://labelstud.io/templates/time_series.html#Output-format-example)

#### Playground
Online sandbox to play around.
[https://labelstud.io/playground/](https://labelstud.io/playground/)

#### Conclusion

> In general the whole interface, with users and tasks make it really interesting to work with and to organize it when having multiple users.

Reviewer: ```suggestion
In general the interface with users and tasks makes it really interesting to work with. It's rather easy to organise the labeling effort when there's multiple users involved.
```

> The tool also has a some flexibility regarding backend data storage as well as for importing data from different sources.

Reviewer: Didn't you mention this before already? About Postgres/SQLite?

Author: yes, I feel it is still valid in the conclusion?

[https://labelstud.io/](https://labelstud.io/)

### Grafana

| Variable      | Value                    |
|---------------|--------------------------|
| Tech          | -                        |
| Maintained    | ✔                        |
| Multivariate  | many plotted in parallel |
| GCP connector | bigquery plugin          |

> The analytics platform for all your metrics

#### Tool

> Grafana is a very nice and flexible tool to build dashboard of all kinds of time series data.

Reviewer: ```suggestion
Grafana is a user-friendly and flexible tool to build dashboards on top of all kinds of time series.
```

#### Setup
For more information: 
[https://grafana.com/docs/grafana/latest/installation/](https://grafana.com/docs/grafana/latest/installation/)

#### Import
Grafana can use many different data sources (officially: [https://grafana.com/docs/grafana/latest/datasources/](https://grafana.com/docs/grafana/latest/datasources/) and many more using plugins: [https://grafana.com/grafana/plugins/?type=datasource](https://grafana.com/grafana/plugins/?type=datasource))).
BigQuery is supported using one of the plugins.

As Grafana is build as a dashboard tool, it is very flexible in making very different plots, combining different parameters.

#### Export
This is one of the less attractive points of Grafana for labelling.
The annotation data is stored in a Grafana database, which can be queried using the API: [https://grafana.com/docs/grafana/latest/http_api/annotations/](https://grafana.com/docs/grafana/latest/http_api/annotations/).

On the other hand annotations can be imported from various sources, so Grafana can be interesting to visualize annotation.

> There is plugin that could store the data in an influxdb. However question is how mature the plugin is ([https://grafana.com/grafana/plugins/novalabs-annotations-panel/](https://grafana.com/grafana/plugins/novalabs-annotations-panel/)).

Reviewer: ```suggestion
There is a plugin that stores the annotations in InfluxDB. However, the question is how mature this [plugin](https://grafana.com/grafana/plugins/novalabs-annotations-panel/) really is.
```

#### Conclusion
Grafana is an excellent tool to visualize time series. 
Annotation is also part of Grafana.
However, when having to label many/long time series the tool is probably limited.
Also, there seems to be a lot of freedom in the tags that are being used. 
This might bring more data garbage into the system.

Interesting tool, but for labeling only to be used for specific use cases.

## So-so tools or rising stars?

### Universal Data Tool

| Variable      | Value                    |
|---------------|--------------------------|
| Tech          | Javascript               |
| Maintained    | ✔                        |
| Multivariate  | many plotted in parallel |
| GCP connector | -                        |

#### Tool

> Universal Data tool is a labelling tool developed for pictures and videos labelling.

Reviewer: ```suggestion
Universal Data Tool is a labelling tool developed to label pictures and videos.
```

Time series have recently been added to the Universal Data Tool, and feels at the time of writing less mature.

#### Data
Time series are not (yet) in the documentation of this tool.

> So the tool might be to immature as of the time of writing.

Reviewer: ```suggestion
For time series, UDT might be too immature right now.
```

So no GCP connector (yet).

#### Conclusion
The tool was built with images and videos in mind. Time series have only recently been added, and the tool feels a bit immature in that perspective.

* [https://github.com/UniversalDataTool/universal-data-tool](https://github.com/UniversalDataTool/universal-data-tool)
* [https://universaldatatool.com/](https://universaldatatool.com/)
* [https://docs.universaldatatool.com/](https://docs.universaldatatool.com/)

### Trainset

| Variable      | Value                      |
| ------------- | -------------------------- |
| Tech          | nodejs                     |
| Maintained    | ✔                         |
| Multivariate  | plot 2 time series simultaneously, annotate 1 |
| GCP connector | No                         |

Trainset is a very lightweight time series annotation tool.

Official page: [https://trainset.geocene.com/](https://trainset.geocene.com/)
Github repo: [https://github.com/geocene/trainset](https://github.com/geocene/trainset) 

#### Tool
[https://github.com/Geocene/trainset](https://github.com/Geocene/trainset) show the install instructions.

On the official page, you can quickly give it a spin with some sample data.

> In the tool, two different time series (a reference series and a labeling series).

Reviewer: ```suggestion
The UI shows 2 different time series (a reference series and a labeling series).
```

### Curve

| Variable      | Value   |
| ------------- | ------- |
| Tech          | python2 |
| Maintained    | No      |
| Multivariate  |         |
| GCP connector |         |

- [https://github.com/baidu/Curve](https://github.com/baidu/Curve)

### TagAnomaly

| Variable      | Value     |
| ------------- | --------- |
| Tech          | R (shiny) |
| Maintained    | ?         |
| Multivariate  |           |
| GCP connector | No        |

> Anomaly detection labeling tool, specifically for multiple time series (one time series per category).

> ￼ Note: This tool was built as a part of a customer engagement, and is not maintained on a regular basis. 

Feels bloated, not so easy to use.

- [https://github.com/Microsoft/TagAnomaly](https://github.com/Microsoft/TagAnomaly)

### CrowdCurio

| Variable      | Value     |
| ------------- | --------- |
| Tech          | Javascript |
| Maintained    | ?         |
| Multivariate  | ?          |
| GCP connector | No        |

Not sure if still alive...

- [https://github.com/CrowdCurio/time-series-annotator](https://github.com/CrowdCurio/time-series-annotator)

## Other tools, possibly interesting

### hastic

> Hastic ([https://hastic.io/](https://hastic.io/) is tool that that allows for easy pattern and anomaly detection.

Reviewer: ```suggestion
Hastic ([https://hastic.io/](https://hastic.io/) is a tool that allows for easy pattern and anomaly detection.
```
```suggestion
Hastic ([https://hastic.io/](https://hastic.io/) is tool that that allows for easy pattern and anomaly detection.
```

There are two parts: hastic server and a hastic grafana app.

So in a sense, this is also a labelling tool, albeit an automated one.

> As it plugs in into Grafana a lot of different data data sources can be used, e.g.

Reviewer: ```suggestion
As it plugs in into Grafana a lot of different data sources can be used, e.g.
```

Reviewer: missing example here?
